modules/imgproc/src/colormap.cpp

```diff
@@ -797,9 +797,29 @@ namespace colormap
             CV_Error(Error::StsAssert, "cv::LUT only supports tables of size 256.");
         if (userColor.type() != CV_8UC1 && userColor.type() != CV_8UC3)
             CV_Error(Error::StsAssert, "cv::LUT only supports tables CV_8UC1 or CV_8UC3.");
-        colormap::UserColorMap cm(userColor.getMat());
-
-        cm(src, dst);
+        if (src.isMat() && userColor.isMat() && ((src.type() == CV_8UC1) || (src.type() == CV_8UC3)))
+        {
+            const cv::Mat srcMat = src.getMat();
+            cv::Mat srcGray;
+            if (srcMat.channels() == 1)
+              srcGray = srcMat;
+            else
+              cv::cvtColor(srcMat, srcGray, cv::COLOR_BGR2GRAY);
+            cv::Mat userColorMat = userColor.getMat();
+            dst.create(src.size(), userColor.type());
+            cv::Mat dstMat = dst.getMat();
+            const size_t elemSize = userColorMat.elemSize();
+            srcGray.forEach<unsigned char>([&](unsigned char& pixel, const int* position) -> void {
+                const int row = position[0];
+                const int col = position[1];
+                memcpy(dstMat.data+row*dstMat.step+col*elemSize, userColorMat.data+pixel*userColorMat.step, elemSize);
```

Comment:
> `+row*dstMat.step+col*elemSize`
> `+pixel*userColorMat.step`

Manual pointer calculations should be avoided.

Reply:
What would be a more canonical way to write that without branching one if() per possible userColorMat.type() ?

Reply:
You have re-implemented LUT() functionality in this forEach block.

In general this not a good approach:
- we have separate code for duplicated functionality.
- code optimization level is not good (memcpy call with non-fixed dynamic elemSize value)

You should use keep using optimized LUT and tune `colormap::UserColorMap` implementation instead.

---

Unfortunately this function is not tested at all:
- there are no accuracy tests
- there are no performance tests.

Reply:
>You have re-implemented LUT() functionality in this forEach block.
Not really. LUT() have the same number of channels for src and dst, while in the context of applyColorMap(), src can be CV_8UC1, and dst CV_8UC3, so this is is more a "fast path" to avoid converting input on the sole purpose of making it "LUT compliant"

>there are no accuracy tests
I do not know how the show that, but I have run tests to check numerical identical results. (basically, not applying the patch, and comparing "gold" results to custom implementation through countNonZerot(absdiff(..))

>there are no performance tests.
I do no know how to add such tests to the OpenCV build/test system, but I have already edited the current PR with a copy/paste of my timings

Reply:
> Not really.

Right. Got it. `cv::LUT()` doesn't support required mode.

---

To get PR merged please move optimization code to the real processing implementation:

> `void ColorMap::operator()(InputArray _src, OutputArray _dst) const`

Next, There are several assumptions about src/dst/userColor mats. Need to add the right checks about these assumptions (code wont process 3D tensors properly). Reuse applicable checks from cv::LUT() implementation.
BTW, prefer to use `CV_Check*()` macros instead of `CV_Assert()`.

---

BTW, probably it makes sense to deprecate and remove later support case with `src.channels = 3`.